tokio-rustls/src/common/mod.rs

```diff
@@ -272,6 +272,14 @@ impl<'a, IO: AsyncRead + AsyncWrite + Unpin, S: Connection> AsyncRead for Stream
                         continue;
                     }
                 }
+                Err(err) if err.kind() == io::ErrorKind::UnexpectedEof => {
+                    self.eof = true;
+                    if prev == buf.remaining() && would_block {
+                        Poll::Ready(Err(err))
+                    } else {
+                        break;
```

Comment:
Should we make sure to yield `UnexpectedEof` on future reads in this case?

Reply:
Yeah, I think so?

Reply:
If we get `UnexpectedEof` from the `reader().read()` call, that means the reader currently has:

* No received plaintext left to read, and
* `!peer_cleanly_closed`, which expands to `!(has_received_close_notify && !self.message_deframer.has_pending())`, and
* `has_seen_eof` on the input stream

`has_seen_eof` should be permanent, but if we have not yet received a `CloseNotify` it is possible that there are still frames pending in the deframer that contain a `CloseNotify`. So we should actually *not* eagerly return `UnexpectedEof` unless the reader gives it to us.

Do you agree?